pkg/kafka/policy.go

```diff
@@ -0,0 +1,196 @@
+	}
+
+	return false
+}
+
+func matchFetchReq(req *proto.FetchReq, rule api.PortRuleKafka) bool {
+	if rule.Topic != "" && !fetchTopicContained(rule.Topic, req.Topics) {
+		return false
+	}
+
+	if rule.ClientID != "" && rule.ClientID != req.ClientID {
+		return false
+	}
+
+	return true
+}
+
+func offsetTopicContained(neededTopic string, topics []proto.OffsetReqTopic) bool {
+	for _, topic := range topics {
+		if topic.Name == neededTopic {
+			return true
+		}
+	}
+
+	return false
+}
+
+func matchOffsetReq(req *proto.OffsetReq, rule api.PortRuleKafka) bool {
+	if rule.Topic != "" && !offsetTopicContained(rule.Topic, req.Topics) {
+		return false
+	}
+
+	if rule.ClientID != "" && rule.ClientID != req.ClientID {
+		return false
+	}
+
+	return true
+}
+
+func topicContained(neededTopic string, topics []string) bool {
+	for _, topic := range topics {
+		if topic == neededTopic {
+			return true
+		}
+	}
+
+	return false
+}
+
+func matchMetadataReq(req *proto.MetadataReq, rule api.PortRuleKafka) bool {
+	if rule.Topic != "" && !topicContained(rule.Topic, req.Topics) {
+		return false
+	}
+
+	if rule.ClientID != "" && rule.ClientID != req.ClientID {
+		return false
+	}
+
+	return true
+}
+
+func offsetCommitTopicContained(neededTopic string, topics []proto.OffsetCommitReqTopic) bool {
+	for _, topic := range topics {
+		if topic.Name == neededTopic {
+			return true
+		}
+	}
+
+	return false
+}
+
+func matchOffsetCommitReq(req *proto.OffsetCommitReq, rule api.PortRuleKafka) bool {
+	if rule.Topic != "" && !offsetCommitTopicContained(rule.Topic, req.Topics) {
+		return false
+	}
+
+	if rule.ClientID != "" && rule.ClientID != req.ClientID {
+		return false
+	}
+
+	return true
+}
+
+func offsetFetchTopicContained(neededTopic string, topics []proto.OffsetFetchReqTopic) bool {
+	for _, topic := range topics {
+		if topic.Name == neededTopic {
+			return true
+		}
+	}
+
+	return false
+}
+
+func matchOffsetFetchReq(req *proto.OffsetFetchReq, rule api.PortRuleKafka) bool {
+	if rule.Topic != "" && !offsetFetchTopicContained(rule.Topic, req.Topics) {
+		return false
+	}
+
+	if rule.ClientID != "" && rule.ClientID != req.ClientID {
+		return false
+	}
+
+	return true
+}
+
+func (req *RequestMessage) ruleMatches(rule api.PortRuleKafka) bool {
+	if rule.GetAPIKey() != 0 && rule.GetAPIKey() != req.kind {
+		return false
+	}
+
+	if rule.GetAPIVersion() != 0 && rule.GetAPIVersion() != req.version {
+		return false
+	}
+
+	switch val := req.request.(type) {
+	case *proto.ProduceReq:
+		return matchProduceReq(val, rule)
+	case *proto.FetchReq:
+		return matchFetchReq(val, rule)
+	case *proto.OffsetReq:
+		return matchOffsetReq(val, rule)
+	case *proto.MetadataReq:
+		return matchMetadataReq(val, rule)
+	case *proto.OffsetCommitReq:
+		return matchOffsetCommitReq(val, rule)
+	case *proto.OffsetFetchReq:
+		return matchOffsetFetchReq(val, rule)
+	case *proto.ConsumerMetadataReq:
+		return true
+	}
+
+	// If all conditions have been met, allow the request
+	return true
```

Comment:
This is basically switch's `default`, shouldn't this be false?

Reply:
Seems fine to me, if there is no rule for request type we should allow it.

Reply:
When all elements of a rule match, then the request is allowed